cmd/thanos/main_test.go

```diff
@@ -0,0 +1,64 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/prober"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/improbable-eng/thanos/pkg/testutil"
+	"github.com/oklog/run"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type TestComponent struct {
+	name string
+}
+
+func (c TestComponent) String() string {
+	return c.name
+}
+
+func queryHTTPGetEndpoint(ctx context.Context, t *testing.T, logger log.Logger, url string) (*http.Response, error) {
+	req, err := http.NewRequest("GET", fmt.Sprintf("http://%s", url), nil)
+	testutil.Ok(t, err)
+	return http.DefaultClient.Do(req.WithContext(ctx))
+}
+
+func TestGenericHttpEndpoints(t *testing.T) {
+	var g run.Group
+	logger := log.NewNopLogger()
+	metricsRegistry := prometheus.NewRegistry()
+	component := TestComponent{name: "sidecar"}
+	ctx := context.Background()
+
+	freePort, err := testutil.FreePort()
+	testutil.Ok(t, err)
+
+	serverAddress := fmt.Sprintf("127.0.0.1:%d", freePort)
+
+	p := prober.NewProber(component, logger)
+	err = defaultHTTPListener(&g, logger, metricsRegistry, serverAddress, p)
+	p.SetReady()
+	testutil.Ok(t, err)
+	go func() { _ = g.Run() }()
+
+	testutil.Ok(t, runutil.Retry(time.Second, ctx.Done(), func() error {
+		resp, err := queryHTTPGetEndpoint(ctx, t, log.NewNopLogger(), path.Join(serverAddress, "/metrics"))
+		testutil.Ok(t, err)
+		testutil.Equals(t, 200, resp.StatusCode)
```

Comment:
there should probably be tests for non-ok responses here as well no?

Reply:
I'm not completely sure for what non-ok responses to check? 
Should I simulate issues with retrieving metrics somehow? 

The tests for Prober statuses are in the Prober package itself if you meant those possibly?

Reply:
I meant a prober that isnt in ready state yet.

Reply:
Ok IIUC you wanted to check how the `defaultHTTPListener` function changes the prober states.

Actually it was ready because I set it for some reason in the test explicitly so I removed it and added tests to verify that the function `defaultHTTPListener` sets it to be healthy but not ready.

PTAL if it's ok with you https://github.com/improbable-eng/thanos/pull/656/commits/b51c5a584295d2b2d6c279ce761d1cd23a8a759a

Reply:
Yes that’s exactly what i  thinking. That lgtm

Reply:
great thanks